File: app/src/main/java/com/selflearning/chemistree/f_games/gamespack/GamesPack.java
package com.selflearning.chemistree.f_games.gamespack;

import androidx.lifecycle.ViewModelProviders;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.selflearning.chemistree.adapter.GamesPackAdapter;
import com.selflearning.chemistree.R;
import com.selflearning.chemistree.games.GameComponentsList;

import java.util.ArrayList;
import java.util.List;

public class GamesPack extends Fragment {

    private Activity activity;
    private Context context;
    private GamesPackAdapter adapter;
    private RecyclerView recyclerView;
    GridLayoutManager gridLayoutManager;
    private List<RowType> rowTypeList = new ArrayList<>();
    private GameComponentsList gameComponentsList;

    private GamesPackViewModel mViewModel;

    public static GamesPack newInstance() {
        return new GamesPack();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.games_pack_fragment, container, false);
        initVar();
        initView(root);
        initListener();
        return root;
    }

    private void initVar(){
        activity = getActivity();
        context = activity.getApplicationContext();
        gridLayoutManager = new GridLayoutManager(context, 2, GridLayoutManager.VERTICAL, false);
        gameComponentsList = new GameComponentsList();


        rowTypeList.add(new DividerRowType("ORGANIC"));

        rowTypeList.add(new GamesRowType(activity, gameComponentsList.getGameComponentsList().get(0)));
        rowTypeList.add(new GamesRowType(activity, gameComponentsList.getGameComponentsList().get(4)));
        rowTypeList.add(new GamesRowType(activity, gameComponentsList.getGameComponentsList().get(5)));
        rowTypeList.add(new GamesRowType(activity, gameComponentsList.getGameComponentsList().get(2)));
        rowTypeList.add(new GamesRowType(activity, gameComponentsList.getGameComponentsList().get(2)));

        rowTypeList.add(new DividerRowType("INORGANIC"));

        rowTypeList.add(new GamesRowType(activity, gameComponentsList.getGameComponentsList().get(0)));
        rowTypeList.add(new GamesRowType(activity, gameComponentsList.getGameComponentsList().get(1)));
        rowTypeList.add(new GamesRowType(activity, gameComponentsList.getGameComponentsList().get(2)));
        rowTypeList.add(new GamesRowType(activity, gameComponentsList.getGameComponentsList().get(3)));
        rowTypeList.add(new GamesRowType(activity, gameComponentsList.getGameComponentsList().get(4)));
        rowTypeList.add(new GamesRowType(activity, gameComponentsList.getGameComponentsList().get(5)));



        gridLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                if(rowTypeList.get(position) instanceof DividerRowType){
                    return 2;
                } else return 1;
            }
        });
    }






    private void initView(View view){
        recyclerView = view.findViewById(R.id.rvGamesPack);
        recyclerView.setLayoutManager(gridLayoutManager);
        recyclerView.setHasFixedSize(true);
        recyclerView.setItemViewCacheSize(10);
        adapter = new GamesPackAdapter(rowTypeList, context);
        recyclerView.setAdapter(adapter);
    }

    private void initListener(){

    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mViewModel = ViewModelProviders.of(this).get(GamesPackViewModel.class);
        // TODO: Use the ViewModel
    }
}

File: app/src/main/java/com/selflearning/chemistree/f_games/gamespack/GamesPack.kt
package com.selflearning.chemistree.f_games.gamespack

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.GridLayoutManager.SpanSizeLookup
import androidx.recyclerview.widget.RecyclerView
import com.selflearning.chemistree.R
import com.selflearning.chemistree.adapter.GamesPackAdapter
import com.selflearning.chemistree.games.GameComponentsList
import java.util.*

class GamesPack : Fragment() {
    private var activity: Activity? = null
    private var adapter: GamesPackAdapter? = null
    private var recyclerView: RecyclerView? = null
    private var gridLayoutManager: GridLayoutManager? = null
    private val rowTypeList: MutableList<RowType> = ArrayList()
    private lateinit var gameComponentsList: GameComponentsList
    private var mViewModel: GamesPackViewModel? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val root = inflater.inflate(R.layout.games_pack_fragment, container, false)
        initVar()
        initView(root)
        initListener()
        return root
    }

    private fun initVar() {
        activity = requireActivity()
        gridLayoutManager = GridLayoutManager(requireContext(), 2, GridLayoutManager.VERTICAL, false)
        gameComponentsList = GameComponentsList()
        rowTypeList.add(DividerRowType("ORGANIC"))
        rowTypeList.add(GamesRowType(requireActivity(), gameComponentsList.gameComponentsList[0]))
        rowTypeList.add(GamesRowType(requireActivity(), gameComponentsList.gameComponentsList[4]))
        rowTypeList.add(GamesRowType(requireActivity(), gameComponentsList.gameComponentsList[5]))
        rowTypeList.add(GamesRowType(requireActivity(), gameComponentsList.gameComponentsList[2]))
        rowTypeList.add(GamesRowType(requireActivity(), gameComponentsList.gameComponentsList[2]))
        rowTypeList.add(DividerRowType("INORGANIC"))
        rowTypeList.add(GamesRowType(requireActivity(), gameComponentsList.gameComponentsList[0]))
        rowTypeList.add(GamesRowType(requireActivity(), gameComponentsList.gameComponentsList[1]))
        rowTypeList.add(GamesRowType(requireActivity(), gameComponentsList.gameComponentsList[2]))
        rowTypeList.add(GamesRowType(requireActivity(), gameComponentsList.gameComponentsList[3]))
        rowTypeList.add(GamesRowType(requireActivity(), gameComponentsList.gameComponentsList[4]))
        rowTypeList.add(GamesRowType(requireActivity(), gameComponentsList.gameComponentsList[5]))
        gridLayoutManager!!.spanSizeLookup = object : SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                return if (rowTypeList[position] is DividerRowType) {
                    2
                } else 1
            }
        }
    }

    private fun initView(view: View) {
        recyclerView = view.findViewById(R.id.rvGamesPack)
        recyclerView?.apply {
            layoutManager = gridLayoutManager
            setHasFixedSize(true)
            setItemViewCacheSize(10)
            adapter = GamesPackAdapter(rowTypeList, requireContext())
            adapter = adapter
        }
    }

    private fun initListener() {}
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        mViewModel = ViewModelProviders.of(this).get(GamesPackViewModel::class.java)
        // TODO: Use the ViewModel
    }

    companion object {
        fun newInstance(): GamesPack {
            return GamesPack()
        }
    }
}